src/preprocessing/target_encoders.rs

```diff
@@ -0,0 +1,233 @@
+/// let oh_vec: Vec<f64> = enc.transform_one(&1).unwrap();
+/// // notice that 1 is actually a zero-th positional label
+/// assert_eq!(oh_vec, vec![1.0, 0.0, 0.0, 0.0, 0.0]);
+/// ```
+///
+/// You can also pass a predefined label enumeration such as a hashmap `HashMap<LabelType, usize>` or a vector `Vec<LabelType>`
+///
+///
+/// ```
+/// use std::collections::HashMap;
+/// use smartcore::preprocessing::target_encoders::OneHotEncoder;
+///
+/// let label_map: HashMap<&str, usize> =
+/// vec![("cat", 2), ("background",0), ("dog", 1)]
+/// .into_iter()
+/// .collect();
+/// let label_vec = vec!["background", "dog", "cat"];
+///
+/// let enc_lv = OneHotEncoder::<&str>::from_positional_label_vec(label_vec);
+/// let enc_lm = OneHotEncoder::<&str>::from_label_map(label_map);
+///
+/// // ["background", "dog", "cat"]
+/// println!("{:?}", enc_lv.get_labels());
+/// assert_eq!(enc_lv.transform_one::<f64>(&"dog"), enc_lm.transform_one::<f64>(&"dog"))
+/// ```
+pub struct OneHotEncoder<LabelType> {
+    label_to_idx: HashMap<LabelType, usize>,
+    labels: Vec<LabelType>,
+    num_classes: usize,
+}
+
+enum LabelDefinition<T> {
+    LabelToClsNumMap(HashMap<T, usize>),
+    PositionalLabel(Vec<T>),
+}
+
+/// Crearte a vector of size num_labels with zeros everywhere and 1 at label_idx (one-hot vector)
+pub fn make_one_hot<T: RealNumber>(label_idx: usize, num_labels: usize) -> Vec<T> {
+    let (pos, neg) = (T::from_f64(1f64).unwrap(), T::from_f64(0f64).unwrap());
+    (0..num_labels)
+        .map(|idx| if idx == label_idx { pos } else { neg })
+        .collect()
+}
+
+impl<'a, LabelType: Hash + Eq + Clone> OneHotEncoder<LabelType> {
+    /// Fit an encoder to a lable list
+    pub fn fit(labels: &[LabelType]) -> Self {
+        let mut label_map: HashMap<LabelType, usize> = HashMap::new();
+        let mut class_num = 0usize;
+        let mut unique_lables: Vec<LabelType> = Vec::new();
+
+        for l in labels {
+            if !label_map.contains_key(&l) {
+                label_map.insert(l.clone(), class_num);
+                unique_lables.push(l.clone());
+                class_num += 1;
+            }
+        }
+        Self {
+            label_to_idx: label_map,
+            num_classes: class_num,
+            labels: unique_lables,
+        }
+    }
+
+    /// Build an encoder from a predefined (label -> class number) map
+    pub fn from_label_map(labels: HashMap<LabelType, usize>) -> Self {
+        Self::from_label_def(LabelDefinition::LabelToClsNumMap(labels))
+    }
+    /// Build an encoder from a predefined positional label-class num vector
+    pub fn from_positional_label_vec(labels: Vec<LabelType>) -> Self {
+        Self::from_label_def(LabelDefinition::PositionalLabel(labels))
+    }
+
+    /// Transform a slice of label types into one-hot vectors
+    /// None is returned if unknown label is encountered
+    pub fn transform<U: RealNumber>(&self, labels: &[LabelType]) -> Vec<Option<Vec<U>>> {
```

Comment:
Can you please rename this method? Method `transform ` is reserved for an operation that 2-dimensional array and transforms every single row using one-hot encoding.
Here is a good example of how method `transform` should look like https://github.com/smartcorelib/smartcore/blob/development/src/decomposition/pca.rs#L126

Reply:
> @VolodymyrOrlov Thank You for taking the time to review this, appreciate the feed-back!
> 
> Let me make sure that I fully understand your comments:
> 
> We do want to have a consistent api with `fit` and `transform` for one-hot encoding, but those must act on entire dataframes as input. My implementation acts on a series (if we stick to the dataframe analogy), so I'll rename those. and make the `fit` and `transform` methods to act on a dataframe.
> 
> I'm a bit confused about what the dataframe should be in this case. smartcore models act on vectors and matricies of `RealNumber` traits. I attempted to replicate Scikit's implementation where by and large you can encode any type of category.
> 
> The simplest solution I can think of is to make `OneHotEncoder` support only `RealNumber` categories and have the user define which ones are categorical.
> What do you think?

@gaxler This is correct, we'd like to support both, series and dataframes. 

I agree with you. `OneHotEncoder` should one-hot-encode not only floats but also other types, like integers and strings. The problem is that current design is not flexible enough to support all these types.

There is an easy and hard paths, depending on how much time you have to work on this feature. 

The easy path is to convert only [nominal and ordinal](https://towardsdatascience.com/data-types-in-statistics-347e152e8bee) floats. In this case you don't have to refactor Smartcore interfaces. You can either infer categorical columns from data (e.g. by taking a random sample and looking at values in the sample), or take a list of column IDs that should be encoded from user, as you've suggested. In fact, I think we should do both.

The hard path is to introduce two new types: dataframe and series. These types will be modeled after Pandas DataFrame and Series and will encapsulate data as a matrix (or as a vector) and additional metadata, like column names and types. Later we can switch all Smartcore algorithms to use these new types instead of matrices and vectors. Since most algorithms cannot work with categorical data directly we'll have to implement internal converters that will extract data from dataframe and series and transform these values into floats or throw an error, if it is not possible.

Let me know what you think about these options. Also, feel free to connect with me in Discord, if you have an account there. My ID is volodymyr.orlov#7062